futures-util/src/async_await/random.rs

```diff
@@ -0,0 +1,51 @@
+use std::{
+    cell::Cell,
+    collections::hash_map::DefaultHasher,
+    hash::Hasher,
+    num::Wrapping,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+#[doc(hidden)]
+pub fn shuffle<T>(slice: &mut [T]) {
+    for i in (1..slice.len()).rev() {
+        slice.swap(i, gen_index(i + 1));
+    }
+}
+
+/// Return a value from `0..n`.
+fn gen_index(n: usize) -> usize {
+    (random() % n as u64) as usize
+}
+
+/// Pseudorandom number generator based on [xorshift*].
+///
+/// [xorshift*]: https://en.wikipedia.org/wiki/Xorshift#xorshift*
+fn random() -> u64 {
+    thread_local! {
+        static RNG: Cell<Wrapping<u64>> = Cell::new(Wrapping(prng_seed()));
+    }
+
+    fn prng_seed() -> u64 {
+        static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+        // Any non-zero seed will do
+        let mut seed = 0;
+        while seed == 0 {
+            let mut hasher = DefaultHasher::new();
+            hasher.write_usize(COUNTER.fetch_add(1, Ordering::Relaxed));
+            seed = hasher.finish();
+        }
```

Comment:
Is there a reason to use an obfuscated deterministic seed here instead of just something like `return 42;`?

Reply:
(Removing the hasher also looks like it would be enough to make this no-std compatible).

Reply:
I guess this makes the seed different per-thread, but so would just incrementing a counter used directly as the seed.

Reply:
There was no particular reason, but I was thinking of using a simpler `cfg(not(feature = "std"))` hasher in no-std.

Reply:
This doesn't seem all that necessary given `xorshift*`s decent statistical randomness, but it doesn't seem like it hurts too much.